### ISSUE TITLE
Compatibility with setuptools w.r.t the `sdist` command

### ISSUE THREAD

**Issue by vadaszd** [warner/python-versioneer#64]
Both ``versioneer`` and ``setuptools`` override the ``sdist`` class of ``distutils.command.sdist``, hard-wiring it as their base-class.
This renders the setuptools 'sdist' command unusable when ``versioneer`` is used with ``setuptools``, as 
 
    setup(
        ...
        cmdclass=versioneer.get_cmdclass(),
        ...
        )
overrides  ``setuptools.command.sdist.sdist`` with ``verisoneer.cmd_sdist``.

One way  to work around this is:

    from setuptools import setup, Extension
    
    # Monkey-patch distutils.command.sdist so that versioneer derives its 
    # command class from setuptools, not distutils
    from setuptools.command.sdist import sdist as setupToolsSdistClass
    import distutils.command.sdist as distutilsSdistModule
    distutilsSdistModule.sdist = setupToolsSdistClass
    # Now we can import verioneer
    import versioneer

Of course it would be much nicer if ``versioneer`` could be told to derive its command objects from the command of another ``distutils`` extension (be that ``setuptools`` or something else). For example ``get_cmdclass()`` could take a dictionary containing command objects to derive from.

**Comment by yvandermeer** [comment]
I ran into this same problem myself, as the "sdist" package generated for my project was missing the package data files and I couldn't figure out why… However, the patch monkey-patch suggested by @vadaszd is not working for me, as it's causing an infinite recursion when the setuptools sdist command calls `make_release_tree()` (which in turn, tries to call the original distutils sdist, which is now itself, causing the recursion).

After some trial and error, I ended up simply manually modifying the `versioneer.py` generated by `versioneer-installer` so that instead of:

    from distutils.command.sdist import sdist as _sdist

it reads:

    from setuptools.command.sdist import sdist as _sdist

Since the generated `versioneer.py` file lives in my project's VCS anyway, this seems like an easy quick-fix. But It does make me wonder how other people using versioneer deal with this issue?

**Comment by warner** [comment]
I'm running into (I think) the same problem when a Versioneer-managed sdist tarball (in this case, `spake2`) is being installed as a dependency of another package (my `magic-wormhole` package). I'm using `setup.py install` or `setup.py develop` on the top-level magic-wormhole, and in the process it runs `setup.py bdist_egg` on spake2 (after first screwing around with the imports table to replace distutils with setuptools). The spake2 egg is not being built correctly: it winds up with the wrong version string (in fact, it gets the same version that the parent magic-wormhole package gets, so something is leaking badly).

I think this is the same issue I was tracking down earlier in #4, so I'm going to investigate it further there.

**Comment by warner** [comment]
I think the current trunk should fix this: I changed it to look for "setuptools" in `sys.modules` to decide which sdist command to create.

(I also fixed the wrong-version problem I had with spake2, which turned out to be a problem with using the `__file__` of `versioneer.py` to locate the project root. The fix was to use the current working directory instead)

I'll close this when I cut the new release.
